AppCleaner: Shizuku throws error while walking file trees

<!-- Thanks for helping me improve SD Maid! Please make sure to fill out as many info's as possible. More info: https://github.com/d4rken-org/sdmaid-se/wiki/Bugs -->

**Describe the bug**
Sdmaid-se 1.2.2-rc0 crashes when using app cleaner. This did not happen with earlier versions. 

**To Reproduce**
Steps to reproduce the behavior:
1. Click on app cleaner scan
4. See error

**Expected behavior**
App cleaner should complete its tasks. 

**Smartphone (please complete the following information):**
 - Device: galaxy s10 plus 
 - OS: Android 12
 - Root status : not rooted. shizuko used
 - SD Maid 2/SE version 1.2.2-rc0 

**Debug log**
If you can reproduce the bug, record the event in [a debug log](https://github.com/d4rken-org/sdmaid-se/wiki/Bugs#debug-log).

**Screenshots/Video**
Screen shot of message after crash included![Screenshot_20240806-184801_SD Maid SE.jpg](https://github.com/user-attachments/assets/f1ddad7a-4570-4ec9-ba21-476095850bd8)



**Additional context**
Add any other context about the problem here.

Thanks. The error happens when reading a directory (from the Shizuku process) and transferring the found file details back (to SD Maids process).

I can't see a specific cause yet.

* Can you check if this folder contains a lot of files/directories: `/storage/7B41-4618/Android/data/com.crittermap.backcountrynavigator.license` ?
* Did it work with v1.2.1? Maybe this is related to the change in 1aa77126 :thinking: 

```java
java.io.IOException: Remote Exception
	at okio.Buffer$inputStream$1.read(SourceFile:26)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:291)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:355)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:181)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:172)
	at java.io.BufferedReader.readLine(BufferedReader.java:335)
	at java.io.BufferedReader.readLine(BufferedReader.java:400)
	at eu.darken.sdmse.common.files.local.ipc.LocalPathLookupIPCFlowKt$toLocalPathLookupFlow$1.invokeSuspend(Unknown Source:107)
	at eu.darken.sdmse.common.files.local.ipc.LocalPathLookupIPCFlowKt$toLocalPathLookupFlow$1.invoke(Unknown Source:12)
	at kotlinx.coroutines.flow.SafeFlow.collectSafely(Unknown Source:2)
	at kotlinx.coroutines.flow.AbstractFlow.collect(Unknown Source:68)
	at kotlinx.coroutines.flow.FlowKt.toCollection(Unknown Source:62)
	at kotlinx.coroutines.flow.FlowKt.toList$default(Unknown Source:5)
	at eu.darken.sdmse.appcleaner.core.scanner.AppScanner.readAppDirs(Unknown Source:604)
	at eu.darken.sdmse.appcleaner.core.scanner.AppScanner$readAppDirs$1.invokeSuspend(Unknown Source:12)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(Unknown Source:8)
	at kotlinx.coroutines.UndispatchedCoroutine.afterResume(Unknown Source:53)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(Unknown Source:22)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(Unknown Source:31)
	at kotlinx.coroutines.DispatchedTask.run(Unknown Source:112)
	at androidx.work.Worker$2.run(Unknown Source:49)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Unknown Source:2)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(Unknown Source:95)
Caused by: android.os.DeadObjectException
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(BinderProxy.java:635)
	at eu.darken.sdmse.common.ipc.RemoteInputStream$Stub$Proxy.readBuffer(Unknown Source:27)
	at okio.Buffer$inputStream$1.read(SourceFile:24)
	... 23 more
```

--- a/app-common-io/src/main/java/eu/darken/sdmse/common/files/local/LocalGateway.kt
+++ b/app-common-io/src/main/java/eu/darken/sdmse/common/files/local/LocalGateway.kt
@@ -34,6 +34,7 @@ import eu.darken.sdmse.common.shizuku.service.runModuleAction
 import eu.darken.sdmse.common.storage.StorageEnvironment
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.onCompletion
 import kotlinx.coroutines.plus
 import kotlinx.coroutines.withContext
 import okio.Sink
@@ -429,7 +430,9 @@ class LocalGateway @Inject constructor(
                 hasRoot() && (mode == Mode.ROOT || !canRead && mode == Mode.AUTO) -> {
                     if (options.isDirect) {
                         log(TAG, VERBOSE) { "walk($mode->ROOT, direct): $path" }
-                        rootOps { it.walk(path, options) }
+                        // We need to keep the resource alive until the caller is done with the Flow
+                        val resource = rootManager.serviceClient.get()
+                        rootOps { it.walk(path, options).onCompletion { resource.close() } }
                     } else {
                         log(TAG, VERBOSE) { "walk($mode->ROOT, indirect): $path" }
                         // Can't pass functions via IPC
@@ -446,7 +449,9 @@ class LocalGateway @Inject constructor(
                 hasShizuku() && (mode == Mode.ADB || !canRead && mode == Mode.AUTO) -> {
                     if (options.isDirect) {
                         log(TAG, VERBOSE) { "walk($mode->ADB, direct): $path" }
-                        adbOps { it.walk(path, options) }
+                        // We need to keep the resource alive until the caller is done with the Flow
+                        val resource = shizukuManager.serviceClient.get()
+                        adbOps { it.walk(path, options).onCompletion { resource.close() } }
                     } else {
                         log(TAG, VERBOSE) { "walk($mode->ADB, indirect): $path" }
                         // Can't pass functions via IPC

